### PR TITLE
MM-29867: Commands to generate test data

### PR DESCRIPTION
**Pull request by agarciamontoro**
#### Summary
This PR adds two new slash commands:
- `/incident test create-incident`, which accepts a playbook ID (that can be chosen from the playbooks the user is a member of, using the autocomplete system), a timestamp and an incident name. It creates an ongoing incident whose creation date is set to the specified timestamp.
- `/incident test bulk-data`, which accepts a number of ongoing incidents, a number of ended incidents, a begin and an end date, and an optional seed. It creates as many ongoing and ended incidents as specified, all of them with their creation date randomly picked between the begin and end dates. The seed, if available, is used to get reproducible results. The names of the incidents are randomly chosen from a list of incident names and a list of fake company names I defined in the code.

Other things I did while I was here:
- I fixed the `sqlstore.NukeDB` function, which failed with a foreign key error when trying to delete `IR_Playbook` before
`IR_PlaybookMember`.
- I moved the `st` command as a subcommand of `/incident test`, for the sake of consistency.
- I excluded `G404: Insecure random number source (rand)` rule from [gosec](https://github.com/securego/gosec): it was complaining because I am using `math/rand` instead of `crpyto/rand` to generate the random numbers needed in the bulk data generation. Of course, we do not need `crypto/rand` for such a task. And, if we ever need to do something crypto-related, I think we will know by ourselves that we need to use `crypto/rand`.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-29867

### CODE REVIEW

**Comment by agarciamontoro** [comment]
E2E tests are failing in CI but not locally. I'll investigate.

File: server/api/playbooks.go [hunk 1]
```diff
@@ -208,6 +209,36 @@ func (h *PlaybookHandler) getPlaybooks(w http.ResponseWriter, r *http.Request) {
 	ReturnJSON(w, playbookResults, http.StatusOK)
 }
 
+func (h *PlaybookHandler) getPlaybooksAutoComplete(w http.ResponseWriter, r *http.Request) {
```
Comment (crspeller) [inline server/api/playbooks.go:212]: I think this needs a permission check.

File: server/incident/incident.go [hunk 2]
```diff
@@ -234,6 +235,9 @@ type Service interface {
 
 	// NukeDB removes all incident related data.
 	NukeDB() error
+
+	// ChangeCreationDate changes the creation date of the specified incident.
+	ChangeCreationDate(incidentID string, creationTimestamp time.Time) error
```
Comment (crspeller) [inline server/incident/incident.go:240]: Seems like a strange function to have. Maybe a function to create with a specific timestamp would make more sense? 0/5 probably fine as is.

**Review by crspeller — CHANGES_REQUESTED** [review]

**Comment by itao** [comment]
@agarciamontoro Could you request review from @muratbayan and @sadohert as they'll be using it in the field?

**Review by itao — APPROVED** [review]

File: server/incident/incident.go (hunk 2 above)
Comment (agarciamontoro) [reply server/incident/incident.go:240]: I'd argue that they are equally weird, actually. The only reason I did it this way was to separate both calls, so there is only one way of creating an incident. As we're both 0/5 on this, I'll leave it as it is just not to add any more work (even if it's negligible).

File: server/api/playbooks.go (hunk 1 above)
Comment (agarciamontoro) [reply server/api/playbooks.go:212]: You're right! Fixed in https://github.com/mattermost/mattermost-plugin-incident-management/pull/358/commits/bc7d032c8d4ea75ae8551aebf22ef19ebbcc95d4

**Comment by agarciamontoro** [comment]
I know now why the E2E tests are failing only in CI (we have an environment variable overriding the server config). I'm waiting on some configuration we need to do in CircleCI itself, but consider it fixed!

**Review by crspeller — APPROVED** [review]

File: server/command/command.go [hunk 3]
```diff
@@ -746,7 +770,7 @@ func (r *Runner) actionSelftest(args []string) {
 
 	if len(args) != 2 || args[0] != confirmPrompt || args[1] != "SELF-TEST" {
 		r.postCommandResponse("Are you sure you want to self-test (which will nuke the database and delete all data -- instances, configuration)? " +
-			"All incident data will be lost. To self-test, type `/incident st CONFIRM SELF-TEST`")
+			"All incident data will be lost. To self-test, type `/incident test self CONFIRM SELF-TEST`")
```
Comment (cpoile) [inline server/command/command.go:773]: maybe:

```suggestion
			"All incident data will be lost. To self-test, type `/incident test self CONFIRM TEST SELF`")
```

along with the change the check would require above

File: server/command/command.go [hunk 4]
```diff
@@ -960,6 +984,348 @@ And... yes, of course, we have emojis
...
 }
 
+func (r *Runner) actionTest(args []string) {
+	if r.pluginAPI.Configuration.GetConfig().ServiceSettings.EnableTesting == nil ||
+		!*r.pluginAPI.Configuration.GetConfig().ServiceSettings.EnableTesting {
+		r.postCommandResponse("Setting `EnableTesting` must be set to `true` to run the test command.")
+		return
+	}
+
+	if !r.pluginAPI.User.HasPermissionTo(r.args.UserId, model.PERMISSION_MANAGE_SYSTEM) {
```
Comment (cpoile) [inline server/command/command.go:994]: Will Murat, Stuart, Paul, etc have system admin privileges when they run the test command?

File: server/command/command.go [hunk 5]
```diff
@@ -960,6 +984,348 @@ And... yes, of course, we have emojis
...
+			return
+		}
+
+		seed = parsedSeed
+	}
+
+	r.generateTestData(ongoing, ended, begin, end, seed)
+}
+
+var fakeCompanyNames = []string{
```
Comment (cpoile) [inline server/command/command.go:1123]: nice. :)

File: server/command/command.go [hunk 6]
```diff
@@ -960,6 +984,348 @@ And... yes, of course, we have emojis
...
+	"Mertz Group",
+	"Welch LLC",
+	"Baumbach Group",
+	"Ward - Schmitt",
+	"Romaguera Group",
+	"Hickle - Kemmer",
+	"Stewart Corp",
+}
+
+var incidentNames = []string{
```
Comment (cpoile) [inline server/command/command.go:1166]: Did you write these? Nice job

**Review by cpoile — APPROVED** [review]
Nice work Alejandro -- this is fiddly stuff to code, but super useful, thanks!

File: server/command/command.go (hunk 4 above)
Comment (agarciamontoro) [reply server/command/command.go:994]: I was assuming that, but maybe it's not true. @muratbayan @sadohert, will you have system admin privileges when running this command to generate the data?

File: server/command/command.go (hunk 6 above)
Comment (agarciamontoro) [reply server/command/command.go:1166]: Yes! It was fun :joy:

File: server/command/command.go (hunk 3 above)
Comment (agarciamontoro) [reply server/command/command.go:773]: It makes it more consistent, thanks for the suggestion!

Comment (agarciamontoro) [reply server/command/command.go]: Done in https://github.com/mattermost/mattermost-plugin-incident-management/pull/358/commits/8e7adeb91db64c6c7da27e29f94ff891bf5bddaf

File: server/command/command.go (hunk 4 above)
Comment (agarciamontoro) [reply server/command/command.go:994]: Confirmed offline that they will have system admin privileges when they run the test command.

**Review by muratbayan — APPROVED** [review]
Reviewed today with Alejandro as I don't have a local setup, the behaviour is great.
Approved

**Comment by agarciamontoro** [comment]
@sadohert are you ok with merging this?

**Comment by sadohert** [comment]
I'm good to merge this.  Functionality looks great!

Probably doesn't need to block this PR, but as discussed in the channel, I was having an issue with a simple `make` of the plugin.  `make deploy` got me up and running in my local dev environment (after make noticed I didn't have my environment details setup correctly and notified me).
image

**Review by sadohert — APPROVED** [review]
I didn't review code.  Only tested the functionality with:
```
/incident test bulk-data 10 3 2020-01-31 2020-11-22 2
```

AND
```
/incident test create-incident 6utgh6qg7p8ndeef9edc583cpc 2020-11-23 PR-Testing
```

And I surfed around to check out what was produced.  Nice job!

**Comment by agarciamontoro** [comment]
@justinegeffen is it ok if I merge this PR? Or do you want to complete the docs before?

**Comment by justinegeffen** [comment]
@agarciamontoro, I am working on the docs at the moment so you can go ahead and merge. :)

**Pull request merged**